File: frontend/src/types/production/board.ts
// ── 工序流程看板（process-board）类型 ──

import type { FieldValue, EquipmentSnapshot } from './execution'

export interface ProcessBoardNode {
  id: string
  node_code: string
  name: string
  stage_name: string
  sort_order: number
}

export interface ProcessBoardPlannedItem {
  batch_id: string
  batch_no: string
  batch_status: string
  plan_order_id: string
  order_no: string
  plan_version: number
  item_id: string
  item_no: number
  planned_quantity: number | null
  unit: string | null
  planned_start: string | null
  planned_end: string | null
  item_status: string
  priority: string
  equipment_id: string | null
}

export interface ProcessBoardExecution {
  execution_id: string
  batch_id: string
  batch_no: string
  execution_seq: number
  /** 锚点执行的状态（进行中/已完成/已中止） */
  status: 'in_progress' | 'completed' | 'aborted'
  /** 看板视角状态：进行中 / 已完成待流转 / 已中止 */
  board_state: 'in_progress' | 'waiting' | 'aborted'
  owner_name: string | null
  started_at: string
  finished_at: string | null
  is_deviation: boolean
  abnormal_count: number
  batch_status: string
  batch_quantity: number | null
  batch_unit: string | null
  equipments: EquipmentSnapshot[]
  field_values: FieldValue[]
  /** 周期监控参考时长（秒）；旧数据或无可用基线时为空 */
  estimated_duration_seconds?: number | null
  /** 按监控基线计算的预计完成时间 */
  expected_finish_at?: string | null
  /** 周期监控状态：monitoring/overdue/not_monitored/sent/resolved/failed */
  timeout_monitor_status?: string | null
  /** 最近一次超时提醒发送时间 */
  timeout_notified_at?: string | null
}

export interface ProcessBoardData {
  route_id: string
  route_name: string
  route_status: string
  nodes: ProcessBoardNode[]
  planned: ProcessBoardPlannedItem[]
  columns: Record<string, ProcessBoardExecution[]>
}

File: backend/app/modules/production/schemas/board.py
"""工序流程看板 API 契约（routes/{route_id}/process-board）。"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.modules.production.schemas.execution import (
    EquipmentSnapshotOut,
    FieldValueOut,
)


class ProcessBoardNodeOut(BaseModel):
    """看板横轴上的一个工序节点。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    node_code: str
    name: str
    stage_name: str
    sort_order: int


class ProcessBoardPlannedItemOut(BaseModel):
    """看板计划批次列：已下达计划单中"已分配未执行"的实际批次及其计划来源。"""

    batch_id: uuid.UUID
    batch_no: str
    batch_status: str
    plan_order_id: uuid.UUID
    order_no: str
    plan_version: int
    item_id: uuid.UUID
    item_no: int
    planned_quantity: float | None
    unit: str | None
    planned_start: datetime | None
    planned_end: datetime | None
    item_status: str
    priority: str
    equipment_id: str | None


class ProcessBoardExecutionOut(BaseModel):
    """工序列中的一条批次（按批次当前位置归入其所在工序）。

    board_state 表示看板视角状态：
    - in_progress：该节点正在执行
    - waiting：该节点已完成、批次等待流转到下一工序（批次整体未完成）
    - aborted：该节点执行被中止
    """

    execution_id: uuid.UUID
    batch_id: uuid.UUID
    batch_no: str
    execution_seq: int
    status: str
    board_state: str
    owner_name: str | None
    started_at: datetime
    finished_at: datetime | None
    is_deviation: bool
    abnormal_count: int
    batch_status: str
    batch_quantity: float | None
    batch_unit: str | None
    equipments: list[EquipmentSnapshotOut] = []
    field_values: list[FieldValueOut] = []


class ProcessBoardOut(BaseModel):
    """工序流程看板整体数据。"""

    route_id: uuid.UUID
    route_name: str
    route_status: str
    nodes: list[ProcessBoardNodeOut]
    planned: list[ProcessBoardPlannedItemOut]
    columns: dict[uuid.UUID, list[ProcessBoardExecutionOut]]
